Pydicer tool class

pchlap opened the pull request:
Hi @dalmouiee, so here is a contribution providing a wrapper class to manage your entire pydicer project. The idea is that a user will create and run the input classes as required. Then once all DICOM is in the PyDicer class will help them orchestrate the rest.

For now, this just provides the basics. But I think we can add a lot more in here. For example functionality to:
- prepare data for nnUNet
- map structures names for consistency
- adding a new data object (such as an auto segmentation). I've got a placeholder for that function in already since I will work on that next.
- loading of images/structures etc...
- and probably lots more....

Curious to hear what you think. PipelineExample.ipynb or the pipeline automated test is probably the best place to see how the pipeline is called using the new PyDicer class :)

dalmouiee commented on pydicer/analyse/data.py:
```diff
@@ -44,53 +44,73 @@ class AnalyseData:
-                }
-                df_rad = pd.read_csv(radiomics_file, index_col=0, dtype=col_types)
-                dfs.append(df_rad)
+        for converted_csv in dataset_directory.glob("*/converted.csv"):
```
I guess the new block of code on line 103 is similar to this, perhaps we can make a function to reduce redundancy?

dalmouiee reviewed (approved):
This is awesome thanks a lot @pchlap! Makes things much easier to handle for sure. Happy for a merge, I just added some comments regarding style but nothing essential :)

pchlap commented:
@fahimirfanalam here's the pull request I was just talking about in case you'd like to take a look :)

pchlap commented on pydicer/cli/run.py:
```diff
@@ -28,7 +27,7 @@
 
 PIPELINE_TOOLS = {
-    "e2e": run_test,
+    # "e2e": run_test, TODO This broke due to some changes. Either we need to fix or remove.
```
Well I deleted the run_test function since that's not really the way its intended to run anymore... If we want the end-to-end test then we can easily put it back it. It's basically equivalent to just pulling the test data and the calling the run_pipeline function :)

pchlap commented on pydicer/dataset/preparation.py:
```diff
@@ -96,12 +96,12 @@ def prepare(self, dataset_name, preparation_function, patients=None, **kwargs):
             object_path = Path(row.path)
 
-            symlink_path = dataset_dir.joinpath(object_path.relative_to(converted_path))
+            symlink_path = dataset_dir.joinpath(object_path.relative_to("data"))
```
Yeah good point, probably best to pull this out as a constant. We basically require the data to be converted into a directory called "data" but maybe one day we want to make this configurable or something...

pchlap commented on pydicer/preprocess/data.py:
```diff
@@ -70,16 +80,19 @@ def preprocess(self):
-            files += list(self.input_directory.glob("**/*"))
+        for directory in input_directory:
+            if config.get_config("enforce_dcm_ext"):
+                files += list(directory.glob("**/*.dcm"))
```
Yep very good point... Again this would probably want to be configurable one day...

pchlap commented on pydicer/analyse/data.py:
```diff
@@ -44,53 +44,73 @@ class AnalyseData:
-                }
-                df_rad = pd.read_csv(radiomics_file, index_col=0, dtype=col_types)
-                dfs.append(df_rad)
+        for converted_csv in dataset_directory.glob("*/converted.csv"):
```
Good find, I've refactored out a function to read the converted data into a Dataframe

pchlap commented on pydicer/visualise/data.py:
```diff
@@ -113,12 +120,25 @@ def visualise(self, patient=None):
                     img_path = Path(img_row.path)
 
-                    img = sitk.ReadImage(str(img_path.joinpath(f"{img_row.modality}.nii.gz")))
+                    # save image inside structure directory
```
I agree for sure, but for the time being I might leave this one. In the future it would be nice to provide the functionality to the user to easily load the data as SimpleITK image objects, but needs a bit more thought around exactly how that should work. But once we do that we can refactor out this sort of code.

pchlap commented:
> This is awesome thanks a lot @pchlap! Makes things much easier to handle for sure. Happy for a merge, I just added some comments regarding style but nothing essential :)

Thanks for the review @dalmouiee, some really useful tips. I've implemented most of them but have left out a few for now to be addressed later. I'll merge this through once everything is passing :)
